### PR TITLE
Ensure the connection used in a failed rollback is discarded

### PR DESCRIPTION
**Pull request by caueguerra**
### Summary

TL;DR: We identified that if a transaction that was rollbacked errors out, it can get re-added to the pool. This will cause the transaction to commit the next time it gets used.

So the connection, on which the transaction was initiated, fails to issues ROLLBACK. Also, the connection is kept alive and is returned to the connection pool. As a consequence of that, if a later transaction comes in, for the same connection, MySQL will COMMIT any pending transaction for that connection before BEGIN. When this happens, it leads to partial data in MySQL and duplicate orders for a checkout.

This can be reproduced by putting a raise inside of rollback_transaction

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 1]
```diff
@@ -317,25 +317,31 @@ def within_new_transaction(isolation: nil, joinable: true)
...
           raise
         ensure
-          if !error && transaction
-            if Thread.current.status == "aborting"
-              rollback_transaction
+          if transaction
+            if error
+              # @connection still holds an open transaction, so we must not
+              # put it back in the pool for reuse
+              @connection.disconnect!  unless transaction.state.rolledback?
```
Comment (jahfer) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:324]: `#discard!`'s documentation says:

> Prevent `@connection`'s finalizer from touching the socket, or otherwise communicating with its server, when it is collected.
>
> Unlike disconnect!, this will not communicate with the server.

Original thinking here was that we don't want to risk not being able to talk to the server _again_ and having this call also 💥. I'm not entirely confident that `#discard!` does everything we want though.

Comment (rafaelfranca) [reply activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:324]: Using `discard!` here will just make this connection be in an undefined state. In other words, a thread can still checkout this connection and any method called on it will raise an exception. What we want is to remove the connection from the connection poll and after that discard it.
